File: src/topsailai_server/agent_daemon/scripts/wait_for_message_result.py
#!/usr/bin/env python3
"""
Wait for Message Result Script

This script sends a new message to a session and then loops waiting for the
processing result by querying messages with processed_msg_id=new_msg_id.

Usage:
    python wait_for_message_result.py --session-id SESSION_ID --message "Hello" --host localhost --port 7373

Environment Variables:
    TOPSAILAI_AGENT_DAEMON_HOST: Server host (default: localhost)
    TOPSAILAI_AGENT_DAEMON_PORT: Server port (default: 7373)
    WAIT_INTERVAL: Polling interval in seconds (default: 2)
    MAX_WAIT_TIME: Maximum wait time in seconds (default: 300)
"""

import argparse
import os
import sys
import time
from typing import Optional, List, Dict, Any

from topsailai_server.agent_daemon import logger
from topsailai_server.agent_daemon.client.message import MessageClient


def wait_for_message_result(
    client: MessageClient,
    session_id: str,
    new_msg_id: str,
    wait_interval: int = 2,
    max_wait_time: int = 300,
    verbose: bool = False
) -> Optional[List[Dict[str, Any]]]:
    """
    Wait for message processing result by polling with processed_msg_id filter.

    Args:
        client: MessageClient instance
        session_id: The session ID
        new_msg_id: The new message ID to wait for
        wait_interval: Polling interval in seconds
        max_wait_time: Maximum wait time in seconds
        verbose: Print verbose output

    Returns:
        List of messages with processed_msg_id=new_msg_id, or None if timeout
    """
    start_time = time.time()
    elapsed = 0

    logger.info("Waiting for message result: session_id=%s, new_msg_id=%s", session_id, new_msg_id)
    print(f"Waiting for message result...")
    print(f"  Session ID: {session_id}")
    print(f"  New Message ID: {new_msg_id}")
    print(f"  Polling interval: {wait_interval}s")
    print(f"  Max wait time: {max_wait_time}s")
    print()

    while elapsed < max_wait_time:
        # Query messages with processed_msg_id=new_msg_id
        messages = client.list_messages(
            session_id=session_id,
            processed_msg_id=new_msg_id,
            sort_key="create_time",
            order_by="asc",
            verbose=False
        )

        if messages:
            logger.info("Found %d message(s) with processed_msg_id=%s", len(messages), new_msg_id)
            print(f"\nFound {len(messages)} message(s) with processed_msg_id={new_msg_id}")
            return messages

        # Print progress
        remaining = max_wait_time - elapsed
        print(f"\rTime elapsed: {int(elapsed)}s / {max_wait_time}s (remaining: {int(remaining)}s)", end="", flush=True)

        # Wait before next poll
        time.sleep(wait_interval)
        elapsed = time.time() - start_time

    # Timeout
    logger.warning("Timeout waiting for message result: session_id=%s, new_msg_id=%s", session_id, new_msg_id)
    print(f"\n\nTimeout! No messages found with processed_msg_id={new_msg_id} after {max_wait_time}s")
    return None


def send_and_wait(
    session_id: str,
    message: str,
    role: str = "user",
    host: str = "localhost",
    port: int = 7373,
    wait_interval: int = 2,
    max_wait_time: int = 300,
    verbose: bool = False
) -> Optional[List[Dict[str, Any]]]:
    """
    Send a message and wait for the processing result.

    Args:
        session_id: The session ID
        message: The message content
        role: Message role (user/assistant)
        host: Server host
        port: Server port
        wait_interval: Polling interval in seconds
        max_wait_time: Maximum wait time in seconds
        verbose: Print verbose output

    Returns:
        List of messages with processed_msg_id=new_msg_id, or None if timeout/error
    """
    client = MessageClient(base_url=f"http://{host}:{port}")

    try:
        # Send the message
        logger.info("Sending message to session %s", session_id)
        print(f"Sending message to session: {session_id}")
        print(f"Message: {message[:100]}{'...' if len(message) > 100 else ''}")

        result = client.send_message(
            session_id=session_id,
            message=message,
            role=role,
            verbose=verbose
        )

        # Extract the new message ID from the result
        # Note: client.send_message() returns the 'data' field from API response,
        # which is already {"msg_id": "..."}, not {"data": {"msg_id": "..."}}
        new_msg_id = None
        if result and isinstance(result, dict):
            new_msg_id = result.get('msg_id')

        if not new_msg_id:
            logger.error("Failed to get new message ID from response")
            print("Error: Failed to get new message ID from response")
            return None

        logger.info("New message created: msg_id=%s", new_msg_id)
        print(f"\nNew message created: {new_msg_id}")
        print("-" * 60)

        # Wait for the result
        return wait_for_message_result(
            client=client,
            session_id=session_id,
            new_msg_id=new_msg_id,
            wait_interval=wait_interval,
            max_wait_time=max_wait_time,
            verbose=verbose
        )

    except Exception as e:
        logger.exception("Error sending message: %s", e)
        print(f"Error: {e}")
        return None


def main():
    parser = argparse.ArgumentParser(
        description='Send a message and wait for the processing result',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  # Send a message and wait for result
  python wait_for_message_result.py --session-id test-session --message "Hello world"

  # With custom host and port
  python wait_for_message_result.py --session-id test-session --message "Hello" --host localhost --port 7373

  # With verbose output
  python wait_for_message_result.py --session-id test-session --message "Hello" --verbose

  # With custom polling interval and timeout
  python wait_for_message_result.py --session-id test-session --message "Hello" --wait-interval 5 --max-wait-time 600
        """
    )

    parser.add_argument('--session-id', type=str, required=True,
                        help='Session ID to send message to')
    parser.add_argument('--message', type=str, required=True,
                        help='Message content')
    parser.add_argument('--role', type=str, default='user',
                        choices=['user', 'assistant'],
                        help='Message role (default: user)')
    parser.add_argument('--host', type=str,
                        default=os.environ.get('TOPSAILAI_AGENT_DAEMON_HOST', 'localhost'),
                        help='Server host (default: localhost)')
    parser.add_argument('--port', type=int,
                        default=int(os.environ.get('TOPSAILAI_AGENT_DAEMON_PORT', '7373')),
                        help='Server port (default: 7373)')
    parser.add_argument('--wait-interval', type=int, default=2,
                        help='Polling interval in seconds (default: 2)')
    parser.add_argument('--max-wait-time', type=int, default=300,
                        help='Maximum wait time in seconds (default: 300)')
    parser.add_argument('--verbose', action='store_true',
                        help='Print verbose output')

    args = parser.parse_args()

    # Update host/port from args if not already set
    if args.host == 'localhost' and 'TOPSAILAI_AGENT_DAEMON_HOST' in os.environ:
        args.host = os.environ['TOPSAILAI_AGENT_DAEMON_HOST']
    if args.port == 7373 and 'TOPSAILAI_AGENT_DAEMON_PORT' in os.environ:
        args.port = int(os.environ['TOPSAILAI_AGENT_DAEMON_PORT'])

    # Send message and wait for result
    result = send_and_wait(
        session_id=args.session_id,
        message=args.message,
        role=args.role,
        host=args.host,
        port=args.port,
        wait_interval=args.wait_interval,
        max_wait_time=args.max_wait_time,
        verbose=args.verbose
    )

    if result:
        print("\n" + "=" * 60)
        print("RESULT:")
        print("=" * 60)
        for msg in result:
            print(f"\n[{msg.get('create_time')}] [{msg.get('msg_id')}] [{msg.get('role')}]")
            print(msg.get('message', ''))
            if msg.get('task_id'):
                print(f">>> task_id: {msg.get('task_id')}")
            if msg.get('task_result'):
                print(f">>> task_result:")
                print(msg.get('task_result'))
        print("\n" + "=" * 60)
        sys.exit(0)
    else:
        print("\nNo result received (timeout or error)")
        sys.exit(1)


if __name__ == '__main__':
    main()

File: src/topsailai_server/agent_daemon/client_go/wait_for_message_result.go
package main

import (
	"bytes"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"time"
)

// Config holds all configuration
type Config struct {
	Host         string
	Port         string
	SessionID    string
	Message      string
	Role         string
	WaitInterval int
	MaxWaitTime  int
	ResultOnly   bool
}

func (c *Config) baseURL() string {
	return fmt.Sprintf("http://%s:%s", c.Host, c.Port)
}

// Message represents a message from the API
type Message struct {
	MsgID          string `json:"msg_id"`
	SessionID      string `json:"session_id"`
	ProcessedMsgID string `json:"processed_msg_id"`
	Role           string `json:"role"`
	Message        string `json:"message"`
	TaskID         string `json:"task_id"`
	TaskResult     string `json:"task_result"`
	CreateTime     string `json:"create_time"`
}

// APIResponse is the standard API response
type APIResponse struct {
	Code    int             `json:"code"`
	Data    json.RawMessage `json:"data"`
	Message string          `json:"message"`
}

// ReceiveMessageRequest is the request body for receiving a message
type ReceiveMessageRequest struct {
	Message        string `json:"message"`
	SessionID      string `json:"session_id"`
	Role           string `json:"role"`
	ProcessedMsgID string `json:"processed_msg_id,omitempty"`
}

// loadConfig loads configuration from environment variables and flags
func loadConfig() *Config {
	cfg := &Config{}

	host := getEnv("TOPSAILAI_AGENT_DAEMON_HOST", "localhost")
	port := getEnv("TOPSAILAI_AGENT_DAEMON_PORT", "7373")
	sessionID := getEnv("TOPSAILAI_SESSION_ID", "")
	message := getEnv("TOPSAILAI_MESSAGE", "")
	role := getEnv("TOPSAILAI_MESSAGE_ROLE", "user")
	waitInterval := intEnv("WAIT_INTERVAL", 2)
	maxWaitTime := intEnv("MAX_WAIT_TIME", 300)
	resultOnly := envBool("RESULT_ONLY")

	flag.StringVar(&cfg.Host, "host", host, "Agent daemon host")
	flag.StringVar(&cfg.Port, "port", port, "Agent daemon port")
	flag.StringVar(&cfg.SessionID, "session-id", sessionID, "Session ID")
	flag.StringVar(&cfg.Message, "message", message, "Message content")
	flag.StringVar(&cfg.Role, "role", role, "Message role (user/assistant)")
	flag.IntVar(&cfg.WaitInterval, "wait-interval", waitInterval, "Wait interval in seconds")
	flag.IntVar(&cfg.MaxWaitTime, "max-wait-time", maxWaitTime, "Max wait time in seconds")
	flag.BoolVar(&cfg.ResultOnly, "result-only", resultOnly, "Only output result")

	flag.Parse()
	return cfg
}

func getEnv(key, defaultVal string) string {
	if val := os.Getenv(key); val != "" {
		return val
	}
	return defaultVal
}

func intEnv(key string, defaultVal int) int {
	if val := os.Getenv(key); val != "" {
		var result int
		fmt.Sscanf(val, "%d", &result)
		return result
	}
	return defaultVal
}

func envBool(key string) bool {
	return strings.ToLower(os.Getenv(key)) == "true" || os.Getenv(key) == "1"
}

// receiveMessage sends a message to the API and returns the new message ID
func receiveMessage(cfg *Config, processedMsgID string) (string, error) {
	url := cfg.baseURL() + "/api/v1/message"

	reqBody := ReceiveMessageRequest{
		Message:        cfg.Message,
		SessionID:      cfg.SessionID,
		Role:           cfg.Role,
		ProcessedMsgID: processedMsgID,
	}

	jsonData, err := json.Marshal(reqBody)
	if err != nil {
		return "", fmt.Errorf("failed to marshal request: %w", err)
	}

	resp, err := http.Post(url, "application/json", bytes.NewBuffer(jsonData))
	if err != nil {
		return "", fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("failed to read response: %w", err)
	}

	// Debug: print raw response if non-2xx
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return "", fmt.Errorf("HTTP %d: %s", resp.StatusCode, string(body))
	}

	var apiResp APIResponse
	if err := json.Unmarshal(body, &apiResp); err != nil {
		return "", fmt.Errorf("failed to parse response (raw: %s): %w", string(body), err)
	}

	if apiResp.Code != 0 {
		return "", fmt.Errorf("API error: %s", apiResp.Message)
	}

	// Parse data to get msg_id
	var data map[string]interface{}
	if err := json.Unmarshal(apiResp.Data, &data); err != nil {
		return "", fmt.Errorf("failed to parse data (raw: %s): %w", string(apiResp.Data), err)
	}

	msgID, ok := data["msg_id"].(string)
	if !ok {
		return "", fmt.Errorf("msg_id not found in response")
	}

	return msgID, nil
}

// listMessages retrieves messages after the given message ID
func listMessages(cfg *Config, afterMsgID string) ([]Message, error) {
	url := cfg.baseURL() + "/api/v1/message?session_id=" + cfg.SessionID + "&sort_key=create_time&order_by=asc&limit=100"

	resp, err := http.Get(url)
	if err != nil {
		return nil, fmt.Errorf("failed to get messages: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	// Debug: print raw response if non-2xx
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("HTTP %d: %s", resp.StatusCode, string(body))
	}

	var apiResp APIResponse
	if err := json.Unmarshal(body, &apiResp); err != nil {
		return nil, fmt.Errorf("failed to parse response (raw: %s): %w", string(body), err)
	}

	if apiResp.Code != 0 {
		return nil, fmt.Errorf("API error: %s", apiResp.Message)
	}

	var messages []Message
	if err := json.Unmarshal(apiResp.Data, &messages); err != nil {
		return nil, fmt.Errorf("failed to parse messages: %w", err)
	}

	// Filter messages that come after afterMsgID
	var result []Message
	found := false
	for _, msg := range messages {
		if msg.MsgID == afterMsgID {
			found = true
			continue
		}
		if found {
			result = append(result, msg)
		}
	}

	return result, nil
}

// formatMessage formats a single message for output
func formatMessage(msg Message) string {
	var sb strings.Builder
	sb.WriteString(fmt.Sprintf("[%s] [%s] [%s]\n", msg.CreateTime, msg.MsgID, msg.Role))
	sb.WriteString(msg.Message)
	if msg.TaskID != "" {
		sb.WriteString(fmt.Sprintf("\n>>> task_id: %s", msg.TaskID))
	}
	if msg.TaskResult != "" {
		sb.WriteString(fmt.Sprintf("\n>>> task_result: %s", msg.TaskResult))
	}
	return sb.String()
}

func main() {
	cfg := loadConfig()

	if cfg.SessionID == "" {
		fmt.Fprintln(os.Stderr, "Error: session-id is required")
		os.Exit(1)
	}
	if cfg.Message == "" {
		fmt.Fprintln(os.Stderr, "Error: message is required")
		os.Exit(1)
	}

	// Send the message
	newMsgID, err := receiveMessage(cfg, "")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error sending message: %v\n", err)
		os.Exit(1)
	}

	// Output new message ID
	fmt.Printf("new_msg_id: %s\n", newMsgID)

	// Wait for result
	waitInterval := time.Duration(cfg.WaitInterval) * time.Second
	maxWaitTime := time.Duration(cfg.MaxWaitTime) * time.Second
	startTime := time.Now()

	for time.Since(startTime) < maxWaitTime {
		time.Sleep(waitInterval)

		messages, err := listMessages(cfg, newMsgID)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Warning: %v\n", err)
			continue
		}

		if len(messages) > 0 {
			if cfg.ResultOnly {
				// Only output the result (task_result or message)
				for _, msg := range messages {
					if msg.TaskResult != "" {
						fmt.Println(msg.TaskResult)
					} else if msg.Message != "" {
						fmt.Println(msg.Message)
					}
				}
			} else {
				// Output full message format
				for _, msg := range messages {
					fmt.Println("---")
					fmt.Print(formatMessage(msg))
					fmt.Println()
				}
			}
			return
		}
	}

	fmt.Fprintln(os.Stderr, "Error: timeout waiting for result")
	os.Exit(1)
}
